### ISSUE TITLE
[RFC] Support JSON Schema for bindings

### ISSUE THREAD

**Issue by magicmatatjahu** [asyncapi/spec#507]
**Is your feature request related to a problem? Please describe.**

Making review for this PR https://github.com/asyncapi/asyncapi-react/pull/259 I realized that we don't have JSON Schema definitions for officially supported bindings by AsyncAPI. Currently we have only markdown documents like for `http` -> https://github.com/asyncapi/bindings/tree/master/http

Without JSON Schema for bindings we cannot:
- validate bindings object in `parser-js`,
- using binding's model in templates for parsed schemas by `parser-js`,
- render correctly bindings in templates and React component - we still can, but the JSON Schema would make it easier.

> how to validate bindings in parser is a different problem, but I'd like to mention it here.

**Can't it be tackled using specification extensions?** and **Describe the solution you'd like**

We can use extensions to define JSON Schema for each officially supported bindings ([link](https://github.com/asyncapi/asyncapi/blob/master/versions/2.0.0/schema.json#L744)).

I very like this approach (treating bindings as extensions), because then versioning of bindings is not related to AsyncAPI versioning at all - bindings can be versioned next to the official AsyncAPI release without any problems.

Problems which I see with this solution:
- where we should store JSON Schemas? Probably the [`bindings`](https://github.com/asyncapi/bindings) repository is the best place.
- what with [`bindingsObject`](https://github.com/asyncapi/asyncapi/blob/master/versions/2.0.0/schema.json#L744) object? Make references to external JSON Schemas in `bindings` repo or maybe remove whole object and then user should define extension of appropriate binding in spec? TBH I don't know. In my PoV I would like to choose second variant, because I think that AsyncAPI JSON Schema should be as small as possible without reference to external sources, but it could degrade DX. Feel free to share your thoughts about that!

**Describe alternatives you've considered**

If we cannot use extension to define JSON Schema for official supported bindings, then only solution as I see, is describing definition [here](https://github.com/asyncapi/asyncapi/blob/master/versions/2.0.0/schema.json#L744), but I very don't like this.

**Comment by derberg** [comment]
At the moment bindings have versioning independent from spec version, and it is good imho.
we should store schemas in bindings repo so it is easier to binding maintainers to maintain them, and just ref those in [here](https://github.com/asyncapi/asyncapi-node/blob/master/schemas/2.0.0.json) but dereference as part of publishing process because we do not want to do it during parsing.

**Comment by magicmatatjahu** [comment]
> ...and just ref those in here but dereference as part of publishing process because we do not want to do it during parsing.

I agree with that! We can use for that [standalone compilation of](https://ajv.js.org/standalone.html) to generate validation function for our schema to improve performance. At the moment we "compile" (including dereferencing) schema in each execution of the `parse` function ([line](https://github.com/asyncapi/parser-js/blob/master/lib/parser.js#L94)). I don't know how we can speed parsing, but I think that it should be faster :)

**Comment by jonaslagoni** [comment]
>  We can use for that [standalone compilation](https://ajv.js.org/standalone.html) to generate validation function for our schema to improve performance. At the moment we "compile" (including dereferencing) schema in each execution of the `parse` function ([line](https://github.com/asyncapi/parser-js/blob/master/lib/parser.js#L94)). I don't know how we can speed parsing, but I think that it should be faster :)

@magicmatatjahu are you talking about splitting the task of validating bindings into separate standalone components and only validate those bindings in use? Instead of everything at once? 🤔

**Comment by magicmatatjahu** [comment]
@jonaslagoni No. Ajv supports compile the JSON Schema to "standalone" function which can be saved to file and then reuse in JS application. In reference to @derberg comments:

> we should store schemas in bindings repo so it is easier to binding maintainers to maintain them, and just ref those in here but dereference as part of publishing process because we do not want to do it during parsing.

So by this standalone function we can dereference refs for binding's JSON Schemas, compiled it and save to function and use this function in parser. By this operation we can move `@asyncapi/specs` package to `devDependencies` -- currently specs are bundled in parser and we can reduce size but it.

**Comment by derberg** [comment]
I would not worry much about custom bindings, we accept any bindings in the spec, look at IBM or Anypoint (in progress). Json schema for extensions will be solved by extension registry.

My feeling is this is not related to spec at all as json schema is a tool, not the spec. Next step would be to first create those JSON Schemas for bindings, right? the only open problem for me is location. Bindings repo seems best as it will be better to maintain it, but on the other hand, in the asyncapi json schema we should not reference files from `master` but rather a tag. Now question is, I'm not decided, is it ok to always push new tag on bindings repo even if there will be a change just in one binding (bindings are versioned separately). For me this is just a technical tag, not related to specification, so sounds good to me, but correct me if I'm wrong

**Comment by KhudaDad414** [comment]
I want to help implement this feature, As I understand we need to:

1. Create JSON Schema for each binding and store them in [binding](https://github.com/asyncapi/bindings) repo.
      - How are we going to approach this? just look at the [specifications](https://github.com/asyncapi/bindings/blob/master/amqp/README.md) and create a JSON Schema for it?
2. Reference those Schemas [here](https://github.com/asyncapi/asyncapi-node/blob/master/schemas/2.0.0.json).
      - Where exactly?
3. generate standalone functions to validate the bindings against JSON Schemas(that we created at the first stage) and use them in parser.
     - Where, at what repo and when?

Sorry, these questions might have obvious answers that I don't see.

**Comment by derberg** [comment]
> How are we going to approach this? just look at the specifications and create a JSON Schema for it?

yeap 😄 but don't worry, some do not have bindings defined yet.
my only concern is that there are different bindings per different AsyncAPI object, like operation and message for example. So there will be maximum of 4 schemas per protocol. So 🤔 we either have 4 files, with proper pattern in file name, or one custom yaml with 4 schemas in it.

> Where exactly?
Here https://github.com/asyncapi/asyncapi-node/blob/master/schemas/2.0.0.json. For example now we have `kafka: {}` and should have `kafka: '$ref: xxx'`
.But not now, this is next step. We do not want to have url refs, at least I think. 

> Where, at what repo and when?

This one is a next step after the previous one, like a final step on how to at the end enable parser to consume these schemas. 

---

Let's just solve it step by step I think and first figure out how and then create files with JSON schemas

**Comment by derberg** [comment]
Bindings schemas are already merged with https://github.com/asyncapi/bindings/pull/60
Now we need to figure out how to handle it in the parser and the asyncapi json schema https://github.com/asyncapi/parser-js/issues/315

**Comment by jonaslagoni** [comment]
I think it makes sense to move this to https://github.com/asyncapi/spec-json-schemas, right?

**Comment by magicmatatjahu** [comment]
@derberg Could you unstale that issue. We have discussed solution for it :)

**Comment by derberg** [comment]
@magicmatatjahu I suggest we close it

- schemas were created long time ago already by Khuda
- Jonas is not integrating them in the `spec-json-schema`

or we move it to `spec-json-schema` and close once PR from Jonas is merged

**Comment by jonaslagoni** [comment]
Agree this can be closed :)

**Comment by jonaslagoni** [comment]
@derberg can you close this one as complete 🙂

**Comment by derberg** [comment]
closing as merge of https://github.com/asyncapi/spec-json-schemas/pull/208 will solve it definitely
